apps/meteor/app/apps/server/bridges/rooms.ts

```diff
@@ -245,6 +245,60 @@ export class AppRoomBridge extends RoomBridge {
 		return users.map((user: ICoreUser) => userConverter.convertToApp(user));
 	}
 
+	protected async getUnreadByUser(roomId: string, userId: string, options: GetMessagesOptions, appId: string): Promise<Array<IMessageRaw>> {
+		this.orch.debugLog(`The App ${appId} is getting the unread messages for the user: "${userId}" in the room: "${roomId}"`);
+
+		const messageConverter = this.orch.getConverters()?.get('messages');
+		if (!messageConverter) {
+			throw new Error('Message converter not found');
+		}
+
+		const room = await Rooms.findOneById(roomId, { projection: { _id: 1 } });
+
+		if (!room) {
+			throw new Error('Room not found');
+		}
+
+		const subscription = await Subscriptions.findOneByRoomIdAndUserId(roomId, userId, { projection: { ls: 1 } });
+
+		const lastSeen = subscription?.ls;
+		if (!lastSeen) {
+			return [];
+		}
+
+		const sort: Sort = options.sort?.createdAt ? { ts: options.sort.createdAt } : { ts: 1 };
+
+		const cursor = Messages.findVisibleByRoomIdBetweenTimestampsNotContainingTypes(roomId, lastSeen, new Date(), [], {
+			...options,
+			sort,
+		});
+
+		const messages = await cursor.toArray();
+		return Promise.all(messages.map((msg) => messageConverter.convertMessageRaw(msg)));
+	}
+
+	protected async getUserUnreadMessageCount(roomId: string, uid: string, appId: string): Promise<number> {
+		this.orch.debugLog(`The App ${appId} is getting the unread messages count of the room: "${roomId}" for the user: "${uid}"`);
+		const user = await Users.findOneById(uid, { projection: { _id: 1 } });
+		if (!user) {
+			throw new Error('User not found');
+		}
+
+		const room = await Rooms.findOneById(roomId, { projection: { _id: 1 } });
+		if (!room) {
+			throw new Error('Room not found');
+		}
```

Comment:
I wonder the value of these. 

sure these add specific errors, but wouldn't `no relation found between room and user` be sufficient too? let the dev check user/room themselves (maybe they could be doing that already and duplicating).

if the roomid or userid is wrong, there wouldnt be a subscription either. we aren't checking for any sort of permission here.

Reply:
I do think you can just receive a uid+rid and check on the subs collection directly. If no value is there, that means user or room are invalid, or that they have no relation between them, which should be enough

Reply:
Right ..I believe we've made this mistake of wanting and validating everything everywhere at every level - causing mostly performance degradation and unnecessary computation. something we're now trying to revert.

the "no relation" error may not be direct, but should suggest to the developer to do their own checks before if needed. keeping our api lean.